File: go/DeleteNode/DeleteNode.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time -> 0(N) Space -> 0(1)
func deleteNode(node *ListNode) {
	var prev *ListNode
	for node.Next != nil {
		node.Val = node.Next.Val
		prev = node
		node = node.Next
	}
	prev.Next = nil
}

func main() {
	two := &ListNode{Val: 3}
	one := &ListNode{Val: 2, Next: two}
	head := &ListNode{Val: 1, Next: one}

	deleteNode(head)

	for head != nil {
		fmt.Println(head.Val)
		head = head.Next
	}
}

File: python/deleteNode.py
class ListNode:
    def __init__(self, val, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(N) Space -> 0(1)
def deleteNode(node: ListNode):
    prev = None
    while node.next is not None:
        node.val = node.next.val
        prev = node
        node = node.next

    if prev:
        prev.next = None


two = ListNode(val=3)
one = ListNode(val=2, next=two)
head = ListNode(val=1, next=one)

deleteNode(head)

while head is not None:
    print(head.val)
    head = head.next
